src/test/scala/org/camunda/feel/impl/builtin/BuiltinContextFunctionsTest.scala

```diff
@@ -234,6 +236,35 @@ class BuiltinContextFunctionsTest
     )
   }
 
+  it should "handle a lazy value mapper" in {
+    val lazyEngine = FeelEngineBuilder()
+      .withCustomValueMapper(new CustomValueMapper {
+        override def toVal(x: Any, innerValueMapper: Any => Val): Option[Val] = x match {
+          case x: Map[String, Any] =>
+            Some {
+              ValContext(
+                Context.StaticContext(
+                  variables = x, // don't eagerly map inner values
+                )
+              )
+            }
+          case  _ => None // fallback to default
+        }
+
+        override def unpackVal(value: Val, innerValueMapper: Val => Any): Option[Any] = {
+          None // fallback to default
+        }
+      })
+      .build()
+
+    lazyEngine.evaluateExpression(
+      """ context put(vars, ["a", "c"], 3) """,
+      Map("vars" -> Map("a" -> Map("b" -> 1, "c" -> 2)))
+    ) should returnResult(
+      Map("a" -> Map("b" -> 1, "c" -> 3))
+    )
+  }
```

Comment:
:wrench: We could simplify the test case and align with [other test cases](https://github.com/camunda/feel-scala/blob/117e335b579b221c6e3bb1c97c88f509c4b4632c/src/test/scala/org/camunda/feel/impl/builtin/BuiltinContextFunctionsTest.scala#L119-L133).

```suggestion
  it should "override nested context entry from a custom context" in {
    evaluateExpression(
      expression = """ context put(vars, ["a", "c"], 3) """,
      variables = Map(
        "vars" ->
          ValContext(
            new MyCustomContext(
              Map("a" -> Map("b" -> 1, "c" -> 2))
            )
        )
      )
    ) should returnResult(
      Map("a" -> Map("b" -> 1, "c" -> 3))
    )
  }
```

Reply:
❤️ like your suggestion. I tried to make it work with a `CustomContext` as well, but have not been successful (I think I had the `CustomContext` at the outermost to provide a `context` to evaluate instead of plain `variables`)

Still, kudos to you about the ValueMapper design. I think the `Optional` return value in combination with the chain of mappers and the `DefaultMapper` is great and nicely extensible 💪